packages/hydrooj/src/model/task.ts

```diff
@@ -25,37 +26,71 @@ async function getFirst(query: Filter<Task>) {
     return null;
 }
 
-class Consumer {
+export class Consumer {
     consuming: boolean;
+    processing: Set<Task> = new Set();
     running?: any;
+    notify: (res?: any) => void;
+    abort: (res?: any) => void;
 
-    constructor(public filter: any, public func: Function, public destoryOnError = true) {
+    constructor(public filter: any, public func: (t: Task) => Promise<void>, public destroyOnError = true, private concurrency = 1) {
         this.consuming = true;
         this.consume();
-        bus.on('app/exit', this.destory);
+        bus.on('app/exit', this.destroy);
     }
 
     async consume() {
         while (this.consuming) {
             try {
+                if (this.processing.size >= this.concurrency) {
+                    // eslint-disable-next-line no-await-in-loop
+                    await new Promise((resolve, reject) => {
+                        this.notify = resolve;
+                        this.abort = reject;
+                    });
+                    continue;
+                }
                 // eslint-disable-next-line no-await-in-loop
                 const res = await getFirst(this.filter);
-                if (res) {
-                    this.running = res;
-                    // eslint-disable-next-line no-await-in-loop
-                    await this.func(res);
-                    this.running = null;
+                if (!res) {
                     // eslint-disable-next-line no-await-in-loop
-                } else await sleep(1000);
+                    await new Promise((resolve, reject) => {
+                        waiterQueue.add(resolve);
+                        this.notify = resolve;
+                        this.abort = reject;
```

Comment:
It shouldn't throw an error when it's being normally closed

Reply:
I'll remove `abort`.